pkg/apiserver/apiserver.go

```diff
@@ -0,0 +1,117 @@
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/version"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+)
+
+var (
+	groupFactoryRegistry = make(announced.APIGroupFactoryRegistry)
+	registry             = registered.NewOrDie("")
+	Scheme               = runtime.NewScheme()
+	Codecs               = serializer.NewCodecFactory(Scheme)
+)
+
+func init() {
+	admissionv1alpha1.AddToScheme(Scheme)
+
+	// we need to add the options to empty v1
+	// TODO fix the server code to avoid this
+	metav1.AddToGroupVersion(Scheme, schema.GroupVersion{Version: "v1"})
+
+	// TODO: keep the generic API server from wanting this
+	unversioned := schema.GroupVersion{Group: "", Version: "v1"}
+	Scheme.AddUnversionedTypes(unversioned,
+		&metav1.Status{},
+		&metav1.APIVersions{},
+		&metav1.APIGroupList{},
+		&metav1.APIGroup{},
+		&metav1.APIResourceList{},
+	)
+}
+
+type Config struct {
+	GenericConfig *genericapiserver.RecommendedConfig
+	ExtraConfig   ExtraConfig
+}
+
+type ExtraConfig struct {
+}
+
+// NamespaceReservationServer contains state for a Kubernetes cluster master/api server.
+type NamespaceReservationServer struct {
+	GenericAPIServer *genericapiserver.GenericAPIServer
+}
+
+type completedConfig struct {
+	GenericConfig genericapiserver.CompletedConfig
+	ExtraConfig   *ExtraConfig
+}
+
+type CompletedConfig struct {
+	// Embed a private pointer that cannot be instantiated outside of this package.
+	*completedConfig
+}
+
+// Complete fills in any fields not set that are required to have valid data. It's mutating the receiver.
+func (c *Config) Complete() CompletedConfig {
+	completedCfg := completedConfig{
+		c.GenericConfig.Complete(),
+		&c.ExtraConfig,
+	}
+
+	completedCfg.GenericConfig.Version = &version.Info{
+		Major: "1",
+		Minor: "0",
+	}
+
+	return CompletedConfig{&completedCfg}
+}
+
+// New returns a new instance of NamespaceReservationServer from the given config.
+func (c completedConfig) New() (*NamespaceReservationServer, error) {
+	genericServer, err := c.GenericConfig.New("kubernetes-namespace-reservation", genericapiserver.EmptyDelegate) // completion is done in Complete, no need for a second time
+	if err != nil {
+		return nil, err
+	}
+
+	s := &NamespaceReservationServer{
+		GenericAPIServer: genericServer,
+	}
+
+	// TODO we're going to need a later k8s.io/apiserver so that we can get discovery to list a different group version for
+	// our endpoint which we'll use to back some custom storage which will consume the AdmissionReview type and give back the correct response
+	//apiGroupInfo := genericapiserver.NewDefaultAPIGroupInfo(wardle.GroupName, registry, Scheme, metav1.ParameterCodec, Codecs)
+	//apiGroupInfo.GroupMeta.GroupVersion = v1alpha1.SchemeGroupVersion
+	//v1alpha1storage := map[string]rest.Storage{}
+	//apiGroupInfo.VersionedResourcesStorageMap["v1alpha1"] = v1alpha1storage
+	//
+	//if err := s.GenericAPIServer.InstallAPIGroup(&apiGroupInfo); err != nil {
+	//	return nil, err
+	//}
```

Comment:
If this is spinning up an API using Kube plumbing this will be a really useful example for us. I will eventually need to figure out how to do similar for the archivist and was wondering how this is best done to simplify auth etc.

Reply:
> If this is spinning up an API using Kube plumbing this will be a really useful example for us. I will eventually need to figure out how to do similar for the archivist and was wondering how this is best done to simplify auth etc.

It is, though this one will be a snowflake.  See https://github.com/kubernetes/sample-apiserver/ for a repo intentionally built to act as an example.  When I'm done, I'm hoping to have a simple library for building future admission webhooks with six or so functions based on business logic.